draft-yasskin-dispatch-bundled-exchanges.md

```diff
@@ -0,0 +1,675 @@
+
+## Top-level structure {#top-level}
+
+*This section is non-normative.*
+
+The bundle is roughly a CBOR item ({{?I-D.ietf-cbor-7049bis}}) with the
+following CDDL ({{?I-D.ietf-cbor-cddl}}) schema, but bundle parsers are required
+to successfully parse some byte strings that aren't valid CBOR. For example,
+sections might have padding between them, or even overlap, as long as the
+embedded relative offsets cause the parsing algorithms in this specification to
+return data.
+
+~~~~~ cddl
+webbundle = [
+  ; 🌐📦 in UTF-8.
+  magic: h'F0 9F 8C 90 F0 9F 93 A6',
+  section-offsets: bytes .cbor {* ($section-name .within tstr) =>
+                                  [ offset: uint, length: uint] },
+  sections: [* $section ],
+  length: bytes .size 8,  ; Big-endian number of bytes in the bundle.
+]
+
+$section-name /= "index" / "manifest" / "critical" / "responses"
+
+$section /= index / manifest / critical / responses
+
+responses = [*response]
+
+~~~~~
+
+## Load a bundle's metadata {#load-metadata}
+
+A bundle holds a series of sections, which can be accessed randomly using the
+information in the `section-offset` CBOR item:
+
+~~~~~ cddl
+section-offsets = {* tstr => [ offset: uint, length: uint] },
+~~~~~
+
+Offsets in this item are relative to the *end* of the section-offset item.
+
+To implement {{semantics-load-metadata}}, the parser MUST run the following
+steps, taking the `stream` as input.
+
+1. Seek to the start of `stream`.
+
+1. Wait for 10 bytes to be available from `stream`.
+
+1. If the wait fails or if reading 10 bytes from `stream` doesn't return the
+   bytes with hex encoding "84 48 F0 9F 8C 90 F0 9F 93 A6" (the CBOR encoding of
+   the 4-item array initial byte and 8-byte bytestring initial byte, followed by
+   🌐📦 in UTF-8), return an error.
+
+1. Let `sectionOffsetsLength` be the result of getting the length of the CBOR
+   bytestring header from `stream` ({{parse-bytestring}}). If this is an error,
+   return that error.
+
+1. If `sectionOffsetsLength` is TBD or greater, return an error.
+
+1. Wait for `sectionOffsetsLength` bytes to be available from `stream`. If this
+   fails, return an error.
+
+1. Read `sectionOffsetsLength` bytes from `stream`, and let `section-offsets` be
+   the result of parsing one CBOR item ({{parse-cbor}}) from them, matching the
+   section-offsets rule in the CDDL ({{!I-D.ietf-cbor-cddl}}) above. If
+   `section-offsets` is an error, return an error.
+
+1. Let `sectionsStart` be the current offset within `stream`. For example, if
+   `sectionOffsetsLength` were 52, `sectionsStart` would be 64.
+
+1. Let `knownSections` be the subset of the {{section-name-registry}} that this
+   client has implemented.
+
+1. Let `ignoredSections` be an empty set.
+
+1. For each `"name"` key in `section-offsets`, if `"name"`'s specification in
+   `knownSections` says not to process other sections, add those sections' names
+   to `ignoredSections`.
+
+1. Let `metadata` be a struct ({{INFRA}}) with no items.
```

Comment:
The idea with structs is that they are immutable.

Reply:
I think you mean that the set of item names is immutable. It's simple to change this to a map, so I've done that. The set of keys is only not-fixed because I wanted to let other specifications add section names that might define new metadata, but I'm not totally sure that's a good idea, so this might go back to a struct later.